File: src/test/java/fr/adrienbrault/idea/symfony2plugin/tests/codeInspection/command/CommandInvokableReturnValueInspectionTest.java
package fr.adrienbrault.idea.symfony2plugin.tests.codeInspection.command;

import fr.adrienbrault.idea.symfony2plugin.tests.SymfonyLightCodeInsightFixtureTestCase;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 * @see fr.adrienbrault.idea.symfony2plugin.codeInspection.command.CommandInvokableReturnValueInspection
 */
public class CommandInvokableReturnValueInspectionTest extends SymfonyLightCodeInsightFixtureTestCase {
    public void setUp() throws Exception {
        super.setUp();
        myFixture.copyFileToProject("classes.php");
    }

    public String getTestDataPath() {
        return "src/test/java/fr/adrienbrault/idea/symfony2plugin/tests/codeInspection/command/fixtures";
    }

    public void testCommandReturningString() {
        assertLocalInspectionContains("test.php", "<?php\n" +
            "use Symfony\\Component\\Console\\Attribute\\AsCommand;\n" +
            "\n" +
            "#[AsCommand(name: 'app:foobar')]\n" +
            "class FoobarCommand\n" +
            "{\n" +
            "    public function __invoke()\n" +
            "    {\n" +
            "        return '<caret>test';\n" +
            "    }\n" +
            "}",
            "Symfony: Command must return an integer value"
        );
    }

    public void testCommandReturningNull() {
        assertLocalInspectionContains("test.php", "<?php\n" +
            "use Symfony\\Component\\Console\\Attribute\\AsCommand;\n" +
            "\n" +
            "#[AsCommand(name: 'app:foobar')]\n" +
            "class FoobarCommand\n" +
            "{\n" +
            "    public function __invoke()\n" +
            "    {\n" +
            "        return n<caret>ull;\n" +
            "    }\n" +
            "}",
            "Symfony: Command must return an integer value"
        );
    }

    public void testCommandReturningIntegerLiteral() {
        assertLocalInspectionNotContains("test.php", "<?php\n" +
            "use Symfony\\Component\\Console\\Attribute\\AsCommand;\n" +
            "\n" +
            "#[AsCommand(name: 'app:foobar')]\n" +
            "class FoobarCommand\n" +
            "{\n" +
            "    public function __invoke()\n" +
            "    {\n" +
            "        return <caret>0;\n" +
            "    }\n" +
            "}",
            "Symfony: Command must return an integer value"
        );
    }

    public void testCommandReturningIntegerVariable() {
        assertLocalInspectionNotContains("test.php", "<?php\n" +
            "use Symfony\\Component\\Console\\Attribute\\AsCommand;\n" +
            "\n" +
            "#[AsCommand(name: 'app:foobar')]\n" +
            "class FoobarCommand\n" +
            "{\n" +
            "    public function __invoke()\n" +
            "    {\n" +
            "        $exitCode = 0;\n" +
            "        return $exit<caret>Code;\n" +
            "    }\n" +
            "}",
            "Symfony: Command must return an integer value"
        );
    }

    public void testCommandReturningCommandSuccess() {
        assertLocalInspectionNotContains("test.php", "<?php\n" +
            "use Symfony\\Component\\Console\\Attribute\\AsCommand;\n" +
            "use Symfony\\Component\\Console\\Command\\Command;\n" +
            "\n" +
            "#[AsCommand(name: 'app:foobar')]\n" +
            "class FoobarCommand\n" +
            "{\n" +
            "    public function __invoke()\n" +
            "    {\n" +
            "        return Command::SUC<caret>CESS;\n" +
            "    }\n" +
            "}",
            "Symfony: Command must return an integer value"
        );
    }

    public void testCommandWithIntReturnTypeIsSkipped() {
        assertLocalInspectionNotContains("test.php", "<?php\n" +
            "use Symfony\\Component\\Console\\Attribute\\AsCommand;\n" +
            "\n" +
            "#[AsCommand(name: 'app:foobar')]\n" +
            "class FoobarCommand\n" +
            "{\n" +
            "    public function __invoke(): int\n" +
            "    {\n" +
            "        return '<caret>test';\n" +
            "    }\n" +
            "}",
            "Symfony: Command must return an integer value"
        );
    }

    public void testNonCommandClassIsIgnored() {
        assertLocalInspectionNotContains("test.php", "<?php\n" +
            "class FoobarClass\n" +
            "{\n" +
            "    public function __invoke(): string\n" +
            "    {\n" +
            "        return 'te<caret>st';\n" +
            "    }\n" +
            "}",
            "Symfony: Command must return an integer value"
        );
    }

    public void testCommandWithMultipleReturns() {
        assertLocalInspectionContains("test.php", "<?php\n" +
            "use Symfony\\Component\\Console\\Attribute\\AsCommand;\n" +
            "\n" +
            "#[AsCommand(name: 'app:foobar')]\n" +
            "class FoobarCommand\n" +
            "{\n" +
            "    public function __invoke()\n" +
            "    {\n" +
            "        if (true) {\n" +
            "            return 0;\n" +
            "        }\n" +
            "        return 'err<caret>or';\n" +
            "    }\n" +
            "}",
            "Symfony: Command must return an integer value"
        );
    }

    public void testTraditionalCommandWithoutAttributeIsIgnored() {
        assertLocalInspectionNotContains("test.php", "<?php\n" +
            "use Symfony\\Component\\Console\\Command\\Command;\n" +
            "\n" +
            "class FoobarCommand extends Command\n" +
            "{\n" +
            "    public function __invoke()\n" +
            "    {\n" +
            "        return 'inva<caret>lid';\n" +
            "    }\n" +
            "}",
            "Symfony: Command must return an integer value"
        );
    }
}

File: src/test/kotlin/fr/adrienbrault/idea/symfony2plugin/tests/codeInspection/command/CommandInvokableReturnValueInspectionTest.kt
package fr.adrienbrault.idea.symfony2plugin.tests.codeInspection.command

import fr.adrienbrault.idea.symfony2plugin.codeInspection.command.CommandInvokableReturnValueInspection
import fr.adrienbrault.idea.symfony2plugin.tests.SymfonyLightCodeInsightFixtureTestCase

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 * @see CommandInvokableReturnValueInspection
 */
class CommandInvokableReturnValueInspectionTest : SymfonyLightCodeInsightFixtureTestCase() {
    override fun setUp() {
        super.setUp()
        myFixture.copyFileToProject("classes.php")
    }

    override fun getTestDataPath(): String =
        "src/test/kotlin/fr/adrienbrault/idea/symfony2plugin/tests/codeInspection/command/fixtures"

    fun testCommandReturningString() {
        assertLocalInspectionContains(CommandInvokableReturnValueInspection::class.java,
            "test.php", """
            <?php
            use Symfony\Component\Console\Attribute\AsCommand;
            
            #[AsCommand(name: 'app:foobar')]
            class FoobarCommand
            {
                public function __invoke()
                {
                    return '<caret>test';
                }
            }
            """.trimIndent(),
            CommandInvokableReturnValueInspection.MESSAGE
        )
    }

    fun testCommandReturningNull() {
        assertLocalInspectionContains(CommandInvokableReturnValueInspection::class.java,
            "test.php", """
            <?php
            use Symfony\Component\Console\Attribute\AsCommand;
            
            #[AsCommand(name: 'app:foobar')]
            class FoobarCommand
            {
                public function __invoke()
                {
                    return n<caret>ull;
                }
            }
            """.trimIndent(),
            CommandInvokableReturnValueInspection.MESSAGE
        )
    }

    fun testCommandReturningIntegerLiteral() {
        assertLocalInspectionNotContains(CommandInvokableReturnValueInspection::class.java,
            "test.php", """
            <?php
            use Symfony\Component\Console\Attribute\AsCommand;
            
            #[AsCommand(name: 'app:foobar')]
            class FoobarCommand
            {
                public function __invoke()
                {
                    return <caret>0;
                }
            }
            """.trimIndent(),
            CommandInvokableReturnValueInspection.MESSAGE
        )
    }

    fun testCommandReturningIntegerVariable() {
        assertLocalInspectionNotContains(CommandInvokableReturnValueInspection::class.java,
            "test.php", """
            <?php
            use Symfony\Component\Console\Attribute\AsCommand;
            
            #[AsCommand(name: 'app:foobar')]
            class FoobarCommand
            {
                public function __invoke()
                {
                    ${'$'}exitCode = 0;
                    return ${'$'}exit<caret>Code;
                }
            }
            """.trimIndent(),
            CommandInvokableReturnValueInspection.MESSAGE
        )
    }

    fun testCommandReturningCommandSuccess() {
        assertLocalInspectionNotContains(CommandInvokableReturnValueInspection::class.java,
            "test.php", """
            <?php
            use Symfony\Component\Console\Attribute\AsCommand;
            use Symfony\Component\Console\Command\Command;
            
            #[AsCommand(name: 'app:foobar')]
            class FoobarCommand
            {
                public function __invoke()
                {
                    return Command::SUC<caret>CESS;
                }
            }
            """.trimIndent(),
            CommandInvokableReturnValueInspection.MESSAGE
        )
    }

    fun testCommandWithIntReturnTypeIsSkipped() {
        assertLocalInspectionNotContains(CommandInvokableReturnValueInspection::class.java,
            "test.php", """
            <?php
            use Symfony\Component\Console\Attribute\AsCommand;
            
            #[AsCommand(name: 'app:foobar')]
            class FoobarCommand
            {
                public function __invoke(): int
                {
                    return '<caret>test';
                }
            }
            """.trimIndent(),
            CommandInvokableReturnValueInspection.MESSAGE
        )
    }

    fun testNonCommandClassIsIgnored() {
        assertLocalInspectionNotContains(CommandInvokableReturnValueInspection::class.java,
            "test.php", """
            <?php
            class FoobarClass
            {
                public function __invoke(): string
                {
                    return 'te<caret>st';
                }
            }
            """.trimIndent(),
            CommandInvokableReturnValueInspection.MESSAGE
        )
    }

    fun testCommandWithMultipleReturns() {
        assertLocalInspectionContains(CommandInvokableReturnValueInspection::class.java,
            "test.php", """
            <?php
            use Symfony\Component\Console\Attribute\AsCommand;
            
            #[AsCommand(name: 'app:foobar')]
            class FoobarCommand
            {
                public function __invoke()
                {
                    if (true) {
                        return 0;
                    }
                    return 'err<caret>or';
                }
            }
            """.trimIndent(),
            CommandInvokableReturnValueInspection.MESSAGE
        )
    }

    fun testTraditionalCommandWithoutAttributeIsIgnored() {
        assertLocalInspectionNotContains(CommandInvokableReturnValueInspection::class.java,
            "test.php", """
            <?php
            use Symfony\Component\Console\Command\Command;
            
            class FoobarCommand extends Command
            {
                public function __invoke()
                {
                    return 'inva<caret>lid';
                }
            }
            """.trimIndent(),
            CommandInvokableReturnValueInspection.MESSAGE
        )
    }
}
